bindings/rust/s2n-tls/src/renegotiate.rs

```diff
@@ -0,0 +1,622 @@
+    callbacks::with_context,
+    config,
+    connection::Connection,
+    enums::CallbackResult,
+    error::{Error, Fallible, Pollable},
+};
+use std::task::Poll;
+
+/// How to handle a renegotiation request.
+///
+/// See s2n_renegotiate_response in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+#[derive(Debug, PartialEq, Copy, Clone)]
+pub enum RenegotiateResponse {
+    Ignore,
+    Reject,
+    Accept,
+}
+
+impl From<RenegotiateResponse> for s2n_renegotiate_response::Type {
+    fn from(input: RenegotiateResponse) -> s2n_renegotiate_response::Type {
+        match input {
+            RenegotiateResponse::Ignore => s2n_renegotiate_response::RENEGOTIATE_IGNORE,
+            RenegotiateResponse::Reject => s2n_renegotiate_response::RENEGOTIATE_REJECT,
+            RenegotiateResponse::Accept => s2n_renegotiate_response::RENEGOTIATE_ACCEPT,
+        }
+    }
+}
+
+/// A callback that triggers when the server requests renegotiation.
+///
+/// Returning "None" will result in the C callback returning an error,
+/// canceling the connection.
+///
+/// See s2n_renegotiate_request_cb in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+//
+// This method returns Option instead of Result because the callback has no mechanism
+// for surfacing errors to the application, so Result would be somewhat deceptive.
+pub trait RenegotiateCallback: 'static + Send + Sync {
+    fn on_renegotiate_request(
+        &mut self,
+        connection: &mut Connection,
+    ) -> Option<RenegotiateResponse>;
+}
+
+impl RenegotiateCallback for RenegotiateResponse {
+    fn on_renegotiate_request(&mut self, _conn: &mut Connection) -> Option<RenegotiateResponse> {
+        Some(*self)
+    }
+}
+
+impl Connection {
+    /// Reset the connection so that it can be renegotiated.
+    ///
+    /// Returning "None" will result in the C callback returning an error,
+    /// canceling the connection.
+    ///
+    /// See s2n_renegotiate_wipe in [the C API documentation](https://github.com/aws/s2n-tls/blob/main/api/unstable/renegotiate.h).
+    /// The Rust equivalent of the connection-specific methods listed are:
+    ///  - Methods to set the file descriptors: not currently supported by rust bindings
+    ///  - Methods to set the send callback:
+    ///    ([Connection::set_send_callback()], [Connection::set_send_context()])
+    ///  - Methods to set the recv callback:
+    ///    ([Connection::set_receive_callback()], [Connection::set_receive_context()])
+    /// Additionally, the application does not need to re-call:
+    /// - [Connection::set_waker()]
+    pub fn wipe_for_renegotiate(&mut self) -> Result<(), Error> {
+        let waker = self.waker().map(|waker| waker.clone());
+        self.wipe_method(|conn| {
+            let result = unsafe { s2n_renegotiate_wipe(conn.as_ptr()).into_result() };
+            result
+        })?;
+        self.set_waker(waker.as_ref())?;
```

Comment:
Saving the waker made my tests nicer, and I don't think someone would want it wiped for renegotiate?

The caller does have to set any connection-level callbacks like the connection initializer again. I don't think that makes keeping the waker a problem, but I also consistently forget how wakers work.

Reply:
I agree that keeping the waker isn't technically a problem, but it does feel like an odd state for the connection to be in, even if only temporarily.

The waker is only there to allow some async callback to be used (conveniently ignoring #4608 ), so if the async callback isn't on the connection it feels odd that the waker still would be.

Reply:
It's maybe not as odd as you're thinking. At the moment all the async callbacks are set on the config, not on the connection. So they're still set after the wipe. But that's not necessarily going to be true forever.

But consistency is probably better. I'm just not going to do keep the waker. The application will have to reset the waker like they reset everything else.